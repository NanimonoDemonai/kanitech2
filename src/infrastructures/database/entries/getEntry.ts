import { prisma } from "src/infrastructures/database/prisma/client";

export const getEntry = (pid: string) =>
  prisma.entry.findUnique({
    where: { pid },
    include: {
      tags: true,
      latestHistory: {
        include: {
          history: true,
        },
      },
    },
  });

export const getEntryWithHistory = (pid: string) =>
  prisma.entry.findUnique({
    where: { pid },
    include: { tags: true, history: true },
  });

export const getEntries = () =>
  prisma.entry.findMany({ include: { tags: true } });
