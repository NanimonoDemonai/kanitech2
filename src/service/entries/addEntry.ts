import { randomUUID } from "crypto";
import { prisma } from "src/service/prisma/client";

interface Props {
  pid: string;
  pageTitle?: string;
  source: string;
  revision?: string;
}

export const addEntry = async (props: Props) => {
  await prisma.entry.create({
    data: { ...props, revision: props.revision ?? randomUUID() },
  });
};
