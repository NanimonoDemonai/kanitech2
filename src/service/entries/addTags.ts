import { Prisma } from "@prisma/client";
import { prisma } from "src/service/prisma/client";

interface Props {
  pid: string;
  tags: string[];
}

const EntryUpdateValidator = Prisma.validator<Prisma.EntryUpdateInput>();
export const addTags = async ({ pid, tags }: Props) => {
  const updateEntry = EntryUpdateValidator({
    tags: {
      connectOrCreate: tags.map((tagName) => ({
        where: { tagName },
        create: { tagName },
      })),
    },
  });
  await prisma.entry.update({
    where: { pid },
    data: updateEntry,
  });
};
