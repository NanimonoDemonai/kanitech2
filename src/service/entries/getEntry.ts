import { prisma } from "src/service/prisma/client";

export const getEntry = (pid: string) =>
  prisma.entry.findUnique({
    where: { pid },
    include: {
      tags: true,
      latestHistory: {
        include: {
          history: true,
        },
      },
    },
  });

export const getEntryWithHistory = (pid: string) =>
  prisma.entry.findUnique({
    where: { pid },
    include: { tags: true, history: true },
  });
