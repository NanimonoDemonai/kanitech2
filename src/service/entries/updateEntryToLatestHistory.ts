import { prisma } from "src/service/prisma/client";

export const updateEntryToLatestHistory = async (pid: string) => {
  const latestHistory = await prisma.history.findFirst({
    where: { entryPid: pid },
    orderBy: { createdAt: "desc" },
  });

  if (!latestHistory) {
    throw new Error("history must exist");
  }

  await prisma.entry.update({
    where: { pid },
    data: {
      latestHistory: {
        upsert: {
          create: { revision: latestHistory.revision },
          update: { revision: latestHistory.revision },
        },
      },
    },
  });
};
