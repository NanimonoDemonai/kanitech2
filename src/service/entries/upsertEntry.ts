import { Prisma } from "@prisma/client";
import { prisma } from "src/service/prisma/client";

const HistoryValidator =
  Prisma.validator<Prisma.HistoryCreateWithoutEntryInput>();
const EntryInputValidator = Prisma.validator<Prisma.EntryCreateInput>();
const EntryUpdateValidator = Prisma.validator<Prisma.EntryUpdateInput>();

interface Props {
  source: string;
  revision: string;
  createdAt: Date;
  message?: string;
  pid: string;
  pageTitle?: string;
}

export const upsertEntry = async ({
  source,
  revision,
  createdAt,
  message,
  pid,
  pageTitle,
}: Props) => {
  const history = HistoryValidator({
    source,
    revision,
    createdAt,
    message,
  });

  const createEntry = EntryInputValidator({
    pid,
    pageTitle,
    createdAt,
    tags: { create: [] },
    history: {
      create: history,
    },
  });
  const updateEntryHistory = EntryUpdateValidator({
    history: {
      connectOrCreate: {
        where: {
          entryRevision: {
            entryPid: pid,
            revision,
          },
        },
        create: history,
      },
    },
  });

  await prisma.entry.upsert({
    where: {
      pid,
    },
    create: createEntry,
    update: updateEntryHistory,
  });
};
