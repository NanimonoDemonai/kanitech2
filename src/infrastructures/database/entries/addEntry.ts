import { randomUUID } from "crypto";
import { addTags } from "src/infrastructures/database/entries/addTags";
import { updateEntryToLatestHistory } from "src/infrastructures/database/entries/updateEntryToLatestHistory";
import { upsertEntry } from "src/infrastructures/database/entries/upsertEntry";

interface Props {
  pid: string;
  pageTitle?: string;
  source: string;
  message?: string;
  revision?: string;
  createdAt?: Date;
  tags?: string[];
}

export const addEntry = async (props: Props) => {
  const now = new Date();
  const { pid, pageTitle, source, revision, createdAt, tags, message } = {
    ...props,
    createdAt: props.createdAt ?? now,
    revision: props.revision ?? randomUUID(),
  };

  try {
    await upsertEntry({ source, revision, createdAt, message, pid, pageTitle });
    await updateEntryToLatestHistory(pid);
    if (tags) {
      await addTags({ pid, tags });
    }
  } catch (e) {
    console.log(e);
  }
};
