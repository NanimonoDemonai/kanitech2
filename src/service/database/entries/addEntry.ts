import { randomUUID } from "crypto";
import { addTags } from "src/service/database/entries/addTags";
import { updateEntryToLatestHistory } from "src/service/database/entries/updateEntryToLatestHistory";
import { upsertEntry } from "src/service/database/entries/upsertEntry";

interface Props {
  pid: string;
  pageTitle?: string;
  source: string;
  message?: string;
  revision?: string;
  createdAt?: Date;
  tags?: string[];
}

export const addEntry = async (props: Props) => {
  const now = new Date();
  const { pid, pageTitle, source, revision, createdAt, tags, message } = {
    ...props,
    createdAt: props.createdAt ?? now,
    revision: props.revision ?? randomUUID(),
  };

  await upsertEntry({ source, revision, createdAt, message, pid, pageTitle });
  await updateEntryToLatestHistory(pid);
  if (tags) {
    await addTags({ pid, tags });
  }
};
