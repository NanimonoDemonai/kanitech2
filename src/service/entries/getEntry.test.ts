import { getEntry } from "src/service/entries/getEntry";
import { prisma } from "src/service/prisma/client";

describe("getEntry", function () {
  beforeAll(async () => {
    await prisma.entry.deleteMany({});

    await prisma.entry.create({
      data: {
        pid: "aaa",
        source: "test",
        revision: 1,
      },
    });
  });
  it("should getEtnry", async function () {
    const entry = await getEntry("aaa");
    expect(entry).toMatchObject({
      pid: "aaa",
      source: "test",
      revision: 1,
    });
  });
});
