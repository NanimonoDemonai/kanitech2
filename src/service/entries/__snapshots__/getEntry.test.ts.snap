// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEntry should getEtnry 1`] = `
{
  "createdAt": 2022-11-20T15:50:22.140Z,
  "id": 1,
  "pageTitle": "",
  "pid": "aaa",
  "revision": 1,
  "source": "",
  "updatedAt": 2022-11-20T15:50:22.140Z,
}
`;
